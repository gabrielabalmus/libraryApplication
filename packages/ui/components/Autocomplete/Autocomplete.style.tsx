import Autocomplete from "@mui/material/Autocomplete";
import { styled as MuiStyled } from "@mui/material/styles";
import { AutocompleteFieldProps } from "./Autocomplete.type";

export const StandardAutocomplete = MuiStyled(
  Autocomplete
)<AutocompleteFieldProps>(({ width, theme }) =>
  theme.unstable_sx({
    width: "100%",
    maxWidth: { md: width, xs: "100%" },
  })
);
