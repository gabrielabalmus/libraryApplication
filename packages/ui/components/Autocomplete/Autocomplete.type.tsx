import { AutocompleteProps } from "@mui/material/Autocomplete";

export interface AutocompleteFieldProps
  extends AutocompleteProps<AutocompleteOptions, boolean, boolean, boolean> {
  width?: string;
}

export interface AutocompleteContainerProps {
  label?: string;
  onChange: (value: AutocompleteOptions | null) => void;
  value: string;
  options: AutocompleteOptions[];
  errorMessage?: string;
  width?: string;
  placeholder?: string;
  disabled?: boolean;
}

export interface AutocompleteOptions {
  id: string;
  name: string;
}
