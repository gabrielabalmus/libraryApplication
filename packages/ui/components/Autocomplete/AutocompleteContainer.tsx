import TextField from "@mui/material/TextField";
import { StandardAutocomplete } from "./Autocomplete.style";
import { AutocompleteContainerProps } from "./Autocomplete.type";

const AutocompleteContainer: React.FC<AutocompleteContainerProps> = ({
  label,
  value,
  options,
  errorMessage,
  onChange,
  width = "inherit",
  placeholder = "",
  disabled = false,
}) => {
  const handleOnChange = (event: any, value: any) => {
    onChange(value);
  };

  const selectedValue =
    (value && options.find((item) => item.id === value)) || null;

  return (
    <StandardAutocomplete
      disablePortal
      width={width}
      options={options}
      getOptionLabel={(option: any) => option.name}
      value={selectedValue}
      onChange={handleOnChange}
      disabled={disabled}
      renderInput={(params) => (
        <TextField
          {...params}
          label={label}
          placeholder={placeholder}
          variant="standard"
          error={!!errorMessage}
          helperText={errorMessage}
        />
      )}
    />
  );
};

export default AutocompleteContainer;
