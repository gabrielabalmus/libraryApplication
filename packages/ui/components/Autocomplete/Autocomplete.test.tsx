import { fireEvent, render, screen } from "@testing-library/react";
import Autocomplete from ".";
import {
  AutocompleteContainerProps,
  AutocompleteOptions,
} from "./Autocomplete.type";
import { useState } from "react";
import "@testing-library/jest-dom";

const mockChange = jest.fn();

const options: AutocompleteOptions[] = [
  {
    id: "option1",
    name: "Option1",
  },
  {
    id: "option2",
    name: "Option2",
  },
  {
    id: "option3",
    name: "Option3",
  },
];
const renderAutocomplete = ({ ...props }: AutocompleteContainerProps) =>
  render(<Autocomplete {...props} />);

describe("snapshot tests for Autocomplete", () => {
  it("renders correctly", () => {
    const { container } = renderAutocomplete({
      label: "Test1",
      value: "option1",
      onChange: mockChange,
      width: "120px",
      options,
    });

    expect(container).toMatchSnapshot();
  });

  it("renders correctly with error message", () => {
    const { container } = renderAutocomplete({
      label: "Test1",
      value: "option1",
      onChange: mockChange,
      options,
      errorMessage: "Error message",
    });

    expect(container).toMatchSnapshot();
  });

  it("renders correctly with placeholder", () => {
    const { container } = renderAutocomplete({
      label: "Test1",
      value: "option1",
      onChange: mockChange,
      options,
      placeholder: "Autocomplete placeholder",
    });

    expect(container).toMatchSnapshot();
  });
});

const AutocompleteState = (
  other: Omit<AutocompleteContainerProps, "onChange" | "options">
) => {
  const [value, setValue] = useState<string>(other.value);

  return (
    <Autocomplete
      {...other}
      value={value}
      onChange={(data: AutocompleteOptions | null) => setValue(data?.id || "")}
      options={options}
    />
  );
};

const wrapper = (
  other: Omit<AutocompleteContainerProps, "onChange" | "options">
) => render(<AutocompleteState {...other} />);

describe("logic tests for Autocomplete", () => {
  it("takes value correctly", async () => {
    wrapper({ value: "option1", placeholder: "Test placeholder" });

    const autocomplete = screen.getByPlaceholderText("Test placeholder");
    expect(autocomplete).toBeInTheDocument();
    expect(autocomplete).toHaveValue("Option1");

    fireEvent.click(autocomplete);

    const value = "Option";
    fireEvent.change(autocomplete, {
      target: {
        value,
      },
    });

    const optionSelected = screen.getByText(options[1].name);

    fireEvent.click(optionSelected);
    fireEvent.click(document.body);

    expect(autocomplete).toHaveValue(options[1].name);
  });
});
