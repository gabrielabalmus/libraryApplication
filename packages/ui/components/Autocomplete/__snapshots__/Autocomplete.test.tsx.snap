// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests for Autocomplete renders correctly 1`] = `
<div>
  <div
    class="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon css-azr3k4-MuiAutocomplete-root"
    width="120px"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for=":r0:"
        id=":r0:-label"
      >
        Test1
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-l4u8b9-MuiInputBase-root-MuiInput-root"
      >
        <input
          aria-autocomplete="list"
          aria-expanded="false"
          aria-invalid="false"
          autocapitalize="none"
          autocomplete="off"
          class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-1x51dt5-MuiInputBase-input-MuiInput-input"
          id=":r0:"
          placeholder=""
          role="combobox"
          spellcheck="false"
          type="text"
          value="Option1"
        />
        <div
          class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
        >
          <button
            aria-label="Clear"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-clearIndicator css-1glvl0p-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-clearIndicator"
            tabindex="-1"
            title="Clear"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
              data-testid="CloseIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            aria-label="Open"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
            tabindex="-1"
            title="Open"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`snapshot tests for Autocomplete renders correctly with error message 1`] = `
<div>
  <div
    class="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon css-1mxgq3g-MuiAutocomplete-root"
    width="inherit"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-colorPrimary Mui-error MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for=":r2:"
        id=":r2:-label"
      >
        Test1
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary Mui-error MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-l4u8b9-MuiInputBase-root-MuiInput-root"
      >
        <input
          aria-autocomplete="list"
          aria-describedby=":r2:-helper-text"
          aria-expanded="false"
          aria-invalid="true"
          autocapitalize="none"
          autocomplete="off"
          class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-1x51dt5-MuiInputBase-input-MuiInput-input"
          id=":r2:"
          placeholder=""
          role="combobox"
          spellcheck="false"
          type="text"
          value="Option1"
        />
        <div
          class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
        >
          <button
            aria-label="Clear"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-clearIndicator css-1glvl0p-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-clearIndicator"
            tabindex="-1"
            title="Clear"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
              data-testid="CloseIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            aria-label="Open"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
            tabindex="-1"
            title="Open"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <p
        class="MuiFormHelperText-root Mui-error MuiFormHelperText-sizeMedium MuiFormHelperText-filled css-1d1r5q-MuiFormHelperText-root"
        id=":r2:-helper-text"
      >
        Error message
      </p>
    </div>
  </div>
</div>
`;

exports[`snapshot tests for Autocomplete renders correctly with placeholder 1`] = `
<div>
  <div
    class="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon css-1mxgq3g-MuiAutocomplete-root"
    width="inherit"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for=":r4:"
        id=":r4:-label"
      >
        Test1
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-l4u8b9-MuiInputBase-root-MuiInput-root"
      >
        <input
          aria-autocomplete="list"
          aria-expanded="false"
          aria-invalid="false"
          autocapitalize="none"
          autocomplete="off"
          class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-1x51dt5-MuiInputBase-input-MuiInput-input"
          id=":r4:"
          placeholder="Autocomplete placeholder"
          role="combobox"
          spellcheck="false"
          type="text"
          value="Option1"
        />
        <div
          class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
        >
          <button
            aria-label="Clear"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-clearIndicator css-1glvl0p-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-clearIndicator"
            tabindex="-1"
            title="Clear"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
              data-testid="CloseIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            aria-label="Open"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
            tabindex="-1"
            title="Open"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
